FreeRTOS/Test/CMock/tasks/Makefile

```diff
@@ -0,0 +1,181 @@
+		 | directories
+	$(CC) $+ $(LDFLAGS) -o $@ 
+
+.SECONDARY:
+# Mock All preprocessed header files
+$(MOCK_SRC_LIST) : $(CPP_FILES)                                                \
+		   Makefile $(PROJECT_DIR)/$(PROJECT).yml                      \
+		  | directories
+	cd $(@D) && cd .. &&                                                   \
+	    ruby $(CMOCK_EXEC_DIR)/cmock.rb -o$(PROJECT_DIR)/$(PROJECT).yml    \
+	    $(<D)/*
+
+.SECONDARY:
+$(CPP_FILES) : | directories
+	echo "executing this rule"
+	cp $(PROJECT_DIR)/FreeRTOSConfig_$*.h $(SCRATCH_DIR)/$*/include/FreeRTOSConfig.h
+	cp $(PROJECT_DIR)/FreeRTOSConfig_$*.h $(SCRATCH_DIR)/$*/include/FreeRTOSConfig.def
+	sed -i '/#ifndef FREERTOS_CONFIG_H/d' $(SCRATCH_DIR)/$*/include/FreeRTOSConfig.def 
+	sed -i '/#define FREERTOS_CONFIG_H/d' $(SCRATCH_DIR)/$*/include/FreeRTOSConfig.def 
+	sed -i '/#endif/d' $(SCRATCH_DIR)/$*/include/FreeRTOSConfig.def 
+	for h_file in $(MOCK_FILES_FP) ; do                                    \
+		echo "execuring rule ";                                        \
+		unifdef -f $(SCRATCH_DIR)/$*/include/FreeRTOSConfig.def $$h_file \
+		> $(SCRATCH_DIR)/$*/cpp/$$(basename $$h_file) ;                \
+	done
+
+.SECONDARY:
+# Build ALL the mock objects for the config
+# This target is called ONLY once per configuration pattern (%)
+$(MOCK_OBJS) : $(MOCK_SRC_LIST)
+	$(eval INCLUDE = -I$(SCRATCH_DIR)/$*/include)
+	$(eval INCLUDE += -I$(SCRATCH_DIR)/$*/mocks)                           \
+	$(eval INCLUDE += -I$(KERNEL_DIR)/include)                             \
+	for files in $^ ; do                                                   \
+       	    new_name=$${files%.c}.o;                                           \
+	    $(CC) -c $$files $(INCLUDE) $(CFLAGS) -fPIC -DUNITY_EXCLUDE_MATH_H \
+		-DUNITY_EXCLUDE_STDINT_H -DUNITY_OUTPUT_CHAR      -Wno-error   \
+		-o $$new_name ;                                                \
+        done
+#find $(@D)  -type f ! -name 'mock_*' -delete
+
+.SECONDARY:
+# Build test_runner.o from test_runner.c
+$(SCRATCH_DIR)/%/test_runner.o : $(SCRATCH_DIR)/%/test_runner.c                \
+                                 $(SCRATCH_DIR)/%/$(PROJECT).o
+	$(eval INCLUDE = -I$(SCRATCH_DIR)/$*/include)
+	$(eval INCLUDE += -I$(SCRATCH_DIR)/$*/mocks)
+	$(eval INCLUDE += -I$(KERNEL_DIR)/include)
+	$(CC) -c $< $(INCLUDE) $(CPPFLAGS)  $(CFLAGS) -Wno-error -o $@
+
+.SECONDARY:
+# Build tasks_utest.o from tasks_utest_%.c
+$(SCRATCH_DIR)/%/$(PROJECT)_utest.o : $(PROJECT_DIR)/$(PROJECT)_utest_%.c      \
+                                      | directories
+	$(eval INCLUDE = -I$(SCRATCH_DIR)/$*/include)
+	$(eval INCLUDE += -I$(SCRATCH_DIR)/$*/mocks)
+	$(eval INCLUDE += -I$(KERNEL_DIR)/include)
+	$(CC) -c $< $(INCLUDE) $(CPPFLAGS)  $(CFLAGS) -Wno-error -o $@
+
+.SECONDARY:
+# Build tasks.so from tasks.c with the custom FreeRTOSConfig.h and place it in
+# its configuration directory
+$(SCRATCH_DIR)/%/$(PROJECT).o : $(KERNEL_DIR)/$(PROJECT).c | directories
```

Comment:
If you run the pre-processor as a separate step, you'll end up with more branches from included macros. That's especially useful for covering both cases of configASSERTs.

Reply:
what do you mean?
we are not running the preprocessor anymore just using unifdef - in a separate step - with each configuration file

Reply:
Right. In testdir.mk theres a rule to run the pre-processor on the target .c file separately so that it expands macros. This makes llcov show a branch for configASSERT.
```
# preprocess proj files to expand macros for coverage
$(PROJ_DIR)/%.i : $(KERNEL_DIR)/%.c
    mkdir -p $(PROJ_DIR)
    $(CC) -E $< $(CPPFLAGS) $(CFLAGS) -o $@

# compile the project objects with coverage instrumented
$(PROJ_DIR)/%.o : $(PROJ_DIR)/%.i
    $(CC) -c $< $(CPPFLAGS) $(INCLUDE_DIR) $(GCC_COV_OPTS) -o $@
```

Reply:
ahh i see.. done